File: src/podcast/services/episode.py
import logging
import os
import uuid

import httpx
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from podcast.config import settings
from podcast.models import Episode, Job

logger = logging.getLogger(__name__)


async def generate_title_from_topic(topic: str) -> str:
    """Generate a short podcast episode title from the topic using GPT-5-nano."""
    try:
        async with httpx.AsyncClient(
            base_url="https://api.openai.com/v1",
            headers={
                "Authorization": f"Bearer {settings.openai_api_key}",
                "Content-Type": "application/json",
            },
            timeout=httpx.Timeout(30.0, connect=10.0),
        ) as client:
            response = await client.post("/chat/completions", json={
                "model": "gpt-5-nano-2025-08-07",
                "messages": [{
                    "role": "user",
                    "content": f"Generate a short, catchy podcast episode title (max 8 words) for this topic. Output ONLY the title, no quotes or punctuation unless part of the title.\n\nTopic: {topic}",
                }],
                "max_completion_tokens": 100,
                "temperature": 0.7,
            })
            response.raise_for_status()
            data = response.json()
            title = data["choices"][0]["message"]["content"].strip().strip('"\'')
            return title[:200] if title else topic[:100]
    except Exception:
        logger.warning("Failed to generate title from topic, using fallback", exc_info=True)
        return topic[:100] if len(topic) > 100 else topic


async def create_episode(
    db: AsyncSession,
    topic: str,
    title: str | None = None,
    description: str | None = None,
    target_length_minutes: int = 30,
    research_model: str | None = None,
    transcript_model: str | None = None,
) -> Episode:
    """Create a new episode and enqueue the first pipeline job."""
    if not title:
        title = await generate_title_from_topic(topic)

    episode = Episode(
        title=title,
        topic=topic,
        description=description,
        target_length_minutes=target_length_minutes,
        status="pending",
        research_model=research_model,
        transcript_model=transcript_model,
    )
    db.add(episode)
    await db.flush()

    job = Job(episode_id=episode.id, step="research", status="pending")
    db.add(job)

    return episode


async def list_episodes(db: AsyncSession) -> list[Episode]:
    result = await db.execute(
        select(Episode)
        .options(selectinload(Episode.jobs))
        .order_by(Episode.created_at.desc())
    )
    return list(result.scalars().all())


async def get_episode(db: AsyncSession, episode_id: uuid.UUID) -> Episode | None:
    result = await db.execute(
        select(Episode)
        .where(Episode.id == episode_id)
        .options(selectinload(Episode.jobs))
    )
    return result.scalar_one_or_none()


async def delete_episode(db: AsyncSession, episode_id: uuid.UUID) -> bool:
    episode = await get_episode(db, episode_id)
    if not episode:
        return False

    # Clean up audio files (use basename to prevent path traversal)
    if episode.audio_filename:
        safe_name = os.path.basename(episode.audio_filename)
        audio_path = os.path.join(settings.audio_dir, safe_name)
        resolved = os.path.realpath(audio_path)
        if resolved.startswith(os.path.realpath(settings.audio_dir)):
            if os.path.exists(resolved):
                os.remove(resolved)
        else:
            logger.warning("Refusing to delete file outside audio dir: %s", audio_path)

    # Clean up segments directory
    segments_dir = os.path.join(settings.audio_dir, "segments", str(episode.id))
    if os.path.isdir(segments_dir):
        import shutil
        shutil.rmtree(segments_dir)

    await db.delete(episode)
    return True


async def retry_episode(db: AsyncSession, episode_id: uuid.UUID) -> Episode | None:
    """Retry a failed episode from its failed step."""
    episode = await get_episode(db, episode_id)
    if not episode or episode.status != "failed":
        return None

    step = episode.failed_step or "research"
    episode.status = "pending"
    episode.error_message = None
    episode.failed_step = None

    job = Job(episode_id=episode.id, step=step, status="pending")
    db.add(job)

    return episode


async def get_next_episode_number(db: AsyncSession) -> int:
    result = await db.execute(select(func.coalesce(func.max(Episode.episode_number), 0)))
    return result.scalar_one() + 1

File: src/services/episode.ts
import fs from "node:fs";
import path from "node:path";
import { eq, and, sql, desc } from "drizzle-orm";
import { db, type Db } from "../database.js";
import { episodes, jobs } from "../schema.js";
import { settings } from "../config.js";
import { createLogger } from "../log-handler.js";

const log = createLogger("episode");

export async function generateTitleFromTopic(topic: string): Promise<string> {
  try {
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${settings.openaiApiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: "gpt-5-nano-2025-08-07",
        messages: [
          {
            role: "user",
            content: `Generate a short, catchy podcast episode title (max 8 words) for this topic. Output ONLY the title, no quotes or punctuation unless part of the title.\n\nTopic: ${topic}`,
          },
        ],
        max_completion_tokens: 100,
        temperature: 0.7,
      }),
      signal: AbortSignal.timeout(30_000),
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = (await res.json()) as any;
    const title = data.choices[0].message.content
      .trim()
      .replace(/^["']|["']$/g, "");
    return title ? title.slice(0, 200) : topic.slice(0, 100);
  } catch {
    log.warn("Failed to generate title from topic, using fallback");
    return topic.length > 100 ? topic.slice(0, 100) : topic;
  }
}

export async function createEpisode(
  topic: string,
  title: string | null | undefined,
  description: string | null | undefined,
  targetLengthMinutes: number,
  userId: string
) {
  if (!title) {
    title = await generateTitleFromTopic(topic);
  }

  return await db.transaction(async (tx) => {
    const [episode] = await tx
      .insert(episodes)
      .values({
        userId,
        title: title!,
        topic,
        description: description ?? null,
        targetLengthMinutes,
        status: "pending",
      })
      .returning();

    await tx.insert(jobs).values({
      episodeId: episode.id,
      step: "research",
      status: "pending",
    });

    return episode;
  });
}

export async function listEpisodes(userId: string) {
  return db.query.episodes.findMany({
    where: eq(episodes.userId, userId),
    with: { jobs: true },
    orderBy: [desc(episodes.createdAt)],
  });
}

export async function getEpisode(episodeId: string, userId: string) {
  return db.query.episodes.findFirst({
    where: and(eq(episodes.id, episodeId), eq(episodes.userId, userId)),
    with: { jobs: true },
  });
}

export async function deleteEpisode(
  episodeId: string,
  userId: string
): Promise<boolean> {
  const episode = await getEpisode(episodeId, userId);
  if (!episode) return false;

  // Clean up audio files
  const userAudioDir = path.join(settings.audioDir, userId);
  if (episode.audioFilename) {
    try {
      const safeName = path.basename(episode.audioFilename);
      const audioPath = path.join(userAudioDir, safeName);
      if (fs.existsSync(audioPath)) {
        const resolved = fs.realpathSync(audioPath);
        const resolvedDir = fs.realpathSync(userAudioDir);
        if (resolved.startsWith(resolvedDir + path.sep)) {
          fs.unlinkSync(resolved);
        } else {
          log.warn("Refusing to delete file outside audio dir: %s", audioPath);
        }
      }
    } catch (e: any) {
      log.warn("Failed to clean up audio file: %s", e.message);
    }
  }

  // Clean up segments directory
  const segmentsDir = path.join(userAudioDir, "segments", episode.id);
  if (fs.existsSync(segmentsDir)) {
    fs.rmSync(segmentsDir, { recursive: true });
  }

  await db.delete(episodes).where(eq(episodes.id, episodeId));
  return true;
}

export async function retryEpisode(episodeId: string, userId: string) {
  const episode = await getEpisode(episodeId, userId);
  if (!episode || episode.status !== "failed") return null;

  const step = episode.failedStep || "research";

  await db
    .update(episodes)
    .set({ status: "pending", errorMessage: null, failedStep: null })
    .where(eq(episodes.id, episodeId));

  await db.insert(jobs).values({
    episodeId: episode.id,
    step,
    status: "pending",
  });

  return getEpisode(episodeId, userId);
}
